src/transactions/contracts/HostVal.cpp

```diff
@@ -0,0 +1,676 @@
+        {
+            return *p < *otherP;
+        }
+    }
+
+    bool
+    operator()(HostBox const& box) const
+    {
+        return ltOther(box);
+    }
+
+    bool
+    operator()(HostVec const& vec) const
+    {
+        releaseAssert(std::holds_alternative<HostVec>(mOther));
+        auto const& otherVec = std::get<HostVec>(mOther);
+        return std::lexicographical_compare(vec.begin(), vec.end(),
+                                            otherVec.begin(), otherVec.end());
+    }
+
+    bool
+    operator()(HostMap const& map) const
+    {
+        releaseAssert(std::holds_alternative<HostMap>(mOther));
+        auto const& otherMap = std::get<HostMap>(mOther);
+        return std::lexicographical_compare(map.begin(), map.end(),
+                                            otherMap.begin(), otherMap.end());
+    }
+
+    bool
+    operator()(uint64_t const& u) const
+    {
+        return ltOther(u);
+    }
+
+    bool
+    operator()(int64_t const& i) const
+    {
+        return ltOther(i);
+    }
+
+    bool
+    operator()(xdr::xstring<> const& str) const
+    {
+        return ltOther(str);
+    }
+
+    bool
+    operator()(xdr::xvector<uint8_t> const& blob) const
+    {
+        return ltOther(blob);
+    }
+
+    bool
+    operator()(HostBigInt const& big) const
+    {
+        return ltOther(big);
+    }
+
+    bool
+    operator()(HostBigRat const& big) const
+    {
+        return ltOther(big);
+    }
+
+    bool
+    operator()(xdr::pointer<LedgerKey> const& lk) const
+    {
+        return ltOtherPtr(lk);
+    }
+
+    bool
+    operator()(xdr::pointer<Operation> const& op) const
+    {
+        return ltOtherPtr(op);
+    }
+
+    bool
+    operator()(xdr::pointer<OperationResult> const& ores) const
+    {
+        return ltOtherPtr(ores);
+    }
+
+    bool
+    operator()(xdr::pointer<Transaction> const& tx) const
+    {
+        return ltOtherPtr(tx);
+    }
+
+    bool
+    operator()(Asset const& a) const
+    {
+        return ltOther(a);
+    }
+
+    bool
+    operator()(Price const& p) const
+    {
+        return ltOther(p);
+    }
+
+    bool
+    operator()(AccountID const& a) const
+    {
+        return ltOther(a);
+    }
+};
+
+bool
+HostValInContext::operator<(HostValInContext const& other) const
+{
+    ZoneScoped;
+    if (mVal.isU63())
+    {
+        if (other.mVal.isU63())
+        {
+            return mVal.asU63() < other.mVal.asU63();
+        }
+        else
+        {
+            return true;
+        }
+    }
+    else if (other.mVal.isU63())
+    {
+        return false;
+    }
+    else if (mVal.getTag() != other.mVal.getTag())
+    {
+        return mVal.getTag() < other.mVal.getTag();
+    }
+
+    // We have the same tag, examine sub-cases.
+    if (mVal.isU32())
+    {
+        return mVal.asU32() < other.mVal.asU32();
+    }
+    else if (mVal.isI32())
+    {
+        return mVal.asI32() < other.mVal.asI32();
+    }
+    else if (mVal.isVoid())
+    {
+        return false;
```

Comment:
Am I mistaken or does this permit comparing void to bool (they have the same tag)? That seems strange.